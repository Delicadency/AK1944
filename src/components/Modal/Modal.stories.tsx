import { StoryFn, Meta } from "@storybook/react";
import { Modal, ModalProps } from "./Modal";
import { useDisclosure } from "@/hooks/useDisclosure";

export default {
  title: "Components/Modal",
  component: Modal,
  parameters: {
    layout: "centered",
    docs: {
      description: {
        component:
          "A reusable modal component that supports custom content, keyboard navigation, and click-outside behavior.",
      },
    },
  },
  argTypes: {
    isOpen: {
      control: "boolean",
      description: "Controls the visibility of the modal",
    },
    onClose: {
      action: "onClose",
      description: "Callback function to close the modal",
    },
    children: {
      control: "text",
      description: "Content to be rendered inside the modal",
    },
  },
} as Meta;

const InteractiveTemplate: StoryFn<ModalProps> = (args) => {
  const { isOpen, onOpen, onClose } = useDisclosure();

  return (
    <div>
      <button onClick={onOpen}>Open Modal</button>
      <Modal {...args} isOpen={isOpen} onClose={onClose} />
    </div>
  );
};

const StaticTemplate: StoryFn<ModalProps> = (args) => <Modal {...args} />;

export const Interactive = InteractiveTemplate.bind({});
Interactive.args = {
  children: (
    <div className="p-4">
      <h2 className="mb-4 text-xl">Interactive Modal</h2>
      <p>This modal can be opened and closed with state management</p>
    </div>
  ),
};

export const Default = StaticTemplate.bind({});
Default.args = {
  isOpen: true,
  children: (
    <div className="p-4">
      <h2 className="mb-4 text-xl">Default Modal</h2>
      <p>This is a basic modal example</p>
    </div>
  ),
};

export const WithCustomContent = StaticTemplate.bind({});
WithCustomContent.args = {
  isOpen: true,
  children: (
    <div className="p-8">
      <h1 className="mb-4 text-2xl">Custom Content Modal</h1>
      <p className="mb-4">This modal demonstrates custom styling and content</p>
      <button className="bg-blue-500 rounded px-4 py-2 text-white">
        Action Button
      </button>
    </div>
  ),
};

export const LongContent = StaticTemplate.bind({});
LongContent.args = {
  isOpen: true,
  children: (
    <div className="p-4">
      <h2 className="mb-4 text-xl">Scrollable Content</h2>
      {Array(10)
        .fill(null)
        .map((_, i) => (
          <p key={i} className="mb-4">
            Lorem ipsum dolor sit amet, consectetur adipiscing elit. Sed do
            eiusmod tempor incididunt ut labore et dolore magna aliqua.
          </p>
        ))}
    </div>
  ),
};
