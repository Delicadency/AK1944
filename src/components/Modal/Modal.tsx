"use client";

import { useEffect, useCallback, useRef } from "react";
import { ModalCloseIcon } from "@/icons/ModalCloseIcon";
import { cn } from "@/utils";

export type ModalProps = {
  isOpen: boolean;
  onClose: () => void;
  children: React.ReactNode;
  className?: string;
};

export const Modal = ({ children, isOpen, onClose, className }: ModalProps) => {
  const modalRef = useRef<HTMLDivElement | null>(null);
  const lastActiveElement = useRef<HTMLElement | null>(null);

  const handleKeyDown = useCallback(
    (e: KeyboardEvent) => {
      if (e.key === "Escape") {
        onClose();
      }
    },
    [onClose],
  );

  useEffect(() => {
    if (isOpen) {
      lastActiveElement.current = document.activeElement as HTMLElement;

      document.body.style.overflow = "hidden";
      window.addEventListener("keydown", handleKeyDown);

      // Move focus to modal
      if (modalRef.current) {
        modalRef.current.focus();
      }
    } else {
      document.body.style.overflow = "auto";

      // Restore focus to the last active element
      if (lastActiveElement.current) {
        lastActiveElement.current.focus();
      }
    }

    return () => {
      document.body.style.overflow = "auto";
      window.removeEventListener("keydown", handleKeyDown);
    };
  }, [isOpen, handleKeyDown]);

  if (!isOpen) return null;

  return (
    <div
      onClick={onClose}
      className="fixed inset-0 z-50 flex h-screen items-center justify-center bg-slate-900/20 backdrop-blur"
      role="dialog"
      aria-modal="true"
      ref={modalRef}
      tabIndex={-1}
    >
      <div
        onClick={(e) => e.stopPropagation()}
        className={cn(
          "desktop:py-15 relative m-auto flex animate-modalAnimation flex-col items-center rounded bg-greenMain px-3 py-12 text-white desktop:px-10",
          className,
        )}
      >
        <button
          className="absolute right-3 top-3"
          aria-label="Zamknij modal"
          onClick={onClose}
        >
          <ModalCloseIcon className="text-inherit" />
        </button>
        {children}
      </div>
    </div>
  );
};
