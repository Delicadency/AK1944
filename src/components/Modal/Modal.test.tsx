import { render, screen, fireEvent } from "@testing-library/react";
import { Modal } from "./Modal";

describe("Modal component", () => {
  it("should render the modal when open", () => {
    render(
      <Modal isModalOpen={true} setIsModalOpen={() => {}}>
        Test
      </Modal>,
    );
    expect(screen.getByRole("dialog")).toBeInTheDocument();
  });

  it("should not render the modal when closed", () => {
    render(
      <Modal isModalOpen={false} setIsModalOpen={() => {}}>
        Test
      </Modal>,
    );
    expect(screen.queryByRole("dialog")).not.toBeInTheDocument();
  });

  it("should call setIsModalOpen when close button is clicked", () => {
    const setIsModalOpen = jest.fn();
    render(
      <Modal isModalOpen={true} setIsModalOpen={setIsModalOpen}>
        Test
      </Modal>,
    );
    fireEvent.click(screen.getByLabelText("Zamknij modal"));
    expect(setIsModalOpen).toHaveBeenCalledTimes(1);
  });

  it("should call setIsModalOpen when clicking outside the modal content", () => {
    const setIsModalOpen = jest.fn();
    render(
      <Modal isModalOpen={true} setIsModalOpen={setIsModalOpen}>
        Test
      </Modal>,
    );
    fireEvent.click(screen.getByRole("dialog"));
    expect(setIsModalOpen).toHaveBeenCalledTimes(1);
  });

  it("should close the modal when Escape key is pressed", () => {
    const setIsModalOpen = jest.fn();
    render(
      <Modal isModalOpen={true} setIsModalOpen={setIsModalOpen}>
        Test
      </Modal>,
    );
    fireEvent.keyDown(window, { key: "Escape" });
    expect(setIsModalOpen).toHaveBeenCalledWith(false);
  });

  it("should not close the modal when any other key than Escape is pressed", () => {
    const setIsModalOpen = jest.fn();
    render(
      <Modal isModalOpen={true} setIsModalOpen={setIsModalOpen}>
        Test
      </Modal>,
    );
    fireEvent.keyDown(window, { key: "Enter" });
    expect(setIsModalOpen).not.toHaveBeenCalled();
  });

  it("should move focus to the modal when opened", () => {
    render(
      <Modal isModalOpen={true} setIsModalOpen={() => {}}>
        Test
      </Modal>,
    );
    const dialog = screen.getByRole("dialog");
    expect(dialog).toHaveFocus();
  });

  it("should move focus to the modal when opened and return focus to the triggering element when the modal closes", () => {
    const setIsModalOpen = jest.fn();
    const { rerender } = render(
      <>
        <button aria-label="Open modal" onClick={() => setIsModalOpen(true)}>
          Open Modal
        </button>
        <Modal isModalOpen={false} setIsModalOpen={setIsModalOpen}>
          Test
        </Modal>
      </>,
    );
    const openButton = screen.getByRole("button", { name: /open modal/i });
    openButton.focus();
    rerender(
      <>
        <button aria-label="Open modal" onClick={() => setIsModalOpen(true)}>
          Open Modal
        </button>
        <Modal isModalOpen={true} setIsModalOpen={setIsModalOpen}>
          Test
        </Modal>
      </>,
    );
    fireEvent.keyDown(window, { key: "Escape" });
    rerender(
      <>
        <button aria-label="Open modal" onClick={() => setIsModalOpen(true)}>
          Open Modal
        </button>
        <Modal isModalOpen={false} setIsModalOpen={setIsModalOpen}>
          Test
        </Modal>
      </>,
    );
    expect(openButton).toHaveFocus();
  });

  it("should lock scroll when modal is open", () => {
    render(
      <Modal isModalOpen={true} setIsModalOpen={() => {}}>
        Test
      </Modal>,
    );
    expect(document.body.style.overflow).toBe("hidden");
  });

  it("should unlock scroll when modal is closed", () => {
    const { rerender } = render(
      <Modal isModalOpen={true} setIsModalOpen={() => {}}>
        Test
      </Modal>,
    );
    rerender(
      <Modal isModalOpen={false} setIsModalOpen={() => {}}>
        Test
      </Modal>,
    );
    expect(document.body.style.overflow).toBe("auto");
  });

  it("should stop event propagation when clicking inside the modal content", () => {
    const setIsModalOpen = jest.fn();
    render(
      <Modal isModalOpen={true} setIsModalOpen={setIsModalOpen}>
        <div>Content</div>
      </Modal>,
    );
    fireEvent.click(screen.getByText("Content"));
    expect(setIsModalOpen).not.toHaveBeenCalled();
  });

  it("should match snapshot", () => {
    const { asFragment } = render(
      <Modal isModalOpen={true} setIsModalOpen={() => {}}>
        Test
      </Modal>,
    );
    expect(asFragment()).toMatchSnapshot();
  });
});
