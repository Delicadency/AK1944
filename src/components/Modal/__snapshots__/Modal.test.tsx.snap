// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Modal component should match snapshot 1`] = `
<DocumentFragment>
  <div
    aria-modal="true"
    class="fixed inset-0 z-50 flex h-screen items-center justify-center bg-slate-900/20 backdrop-blur"
    role="dialog"
    tabindex="-1"
  >
    <div
      class="relative m-auto flex animate-modalAnimation flex-col items-center rounded bg-greenMain px-3 py-12 text-white desktop:px-10 desktop:py-[60px]"
    >
      <button
        aria-label="Zamknij modal"
        class="absolute right-3 top-3"
      >
        <svg
          fill="none"
          height="44"
          viewBox="0 0 44 44"
          width="44"
          xmlns="http://www.w3.org/2000/svg"
        >
          <path
            d="M29 13H15C13.8954 13 13 13.8954 13 15V29C13 30.1046 13.8954 31 15 31H29C30.1046 31 31 30.1046 31 29V15C31 13.8954 30.1046 13 29 13Z"
            stroke="#FFFCF5"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M19 19L25 25"
            stroke="#FFFCF5"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
          <path
            d="M25 19L19 25"
            stroke="#FFFCF5"
            stroke-linecap="round"
            stroke-linejoin="round"
            stroke-width="2"
          />
        </svg>
      </button>
      Test
    </div>
  </div>
</DocumentFragment>
`;
